File: src/main/java/arc/arc/sync/SkillsSync.java
package arc.arc.sync;

import arc.arc.ARC;
import arc.arc.sync.base.Context;
import arc.arc.sync.base.Sync;
import arc.arc.sync.base.SyncData;
import arc.arc.sync.base.SyncRepo;
import com.google.gson.annotations.SerializedName;
import dev.aurelium.auraskills.api.AuraSkillsApi;
import dev.aurelium.auraskills.api.registry.NamespacedId;
import dev.aurelium.auraskills.api.skill.Skill;
import dev.aurelium.auraskills.api.user.SkillsUser;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.bukkit.Bukkit;
import org.bukkit.scheduler.BukkitRunnable;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicInteger;

public class SkillsSync implements Sync {



    SyncRepo<UserSkillData> syncRepo;
    Map<UUID, Boolean> loaded = new ConcurrentHashMap<>();

    public SkillsSync() {
        this.syncRepo = SyncRepo.builder(UserSkillData.class)
                .key("arc.skills_data")
                .redisManager(ARC.redisManager)
                .dataApplier(this::applySkillData)
                .dataProducer(this::getSkillData)
                .build();
    }

    @Override
    public void playerJoin(UUID uuid) {
        AtomicInteger counter = new AtomicInteger(0);
        new BukkitRunnable() {
            @Override
            public void run() {
                if (Bukkit.getPlayer(uuid) == null) {
                    cancel();
                    return;
                }
                SkillsUser user = AuraSkillsApi.get().getUser(uuid);
                if (user == null) {
                    if (counter.incrementAndGet() > 20) {
                        ARC.warn("PlayerData is null for " + uuid + " for 20 cycles. Cancelling task.");
                        cancel();
                    }
                    return;
                }
                syncRepo.loadAndApplyData(uuid, false);
                loaded.put(uuid, true);
                cancel();
            }
        }.runTaskTimer(ARC.plugin, 5L, 5L);
    }

    @Override
    public void forceSave(UUID uuid) {
        playerQuit(uuid);
    }

    @Override
    public void playerQuit(UUID uuid) {
        if(!loaded.containsKey(uuid)) return;
        Context context = new Context();
        context.put("uuid", uuid);
        syncRepo.saveAndPersistData(context, false);
        loaded.remove(uuid);
    }


    @Data
    @AllArgsConstructor @NoArgsConstructor
    public static class SkillInfo {
        @SerializedName("i")
        String id;
        @SerializedName("l")
        int level;
        @SerializedName("x")
        double xp;
    }


    public UserSkillData getSkillData(Context context) {
        UUID uuid = context.get("uuid");
        Collection<Skill> skills = AuraSkillsApi.get().getGlobalRegistry().getSkills();
        SkillsUser user = AuraSkillsApi.get().getUser(uuid);
        if (user == null) {
            ARC.warn("User with uuid {} not found while getting skill data", uuid);
            return null;
        }

        List<SkillInfo> skillInfoList = new ArrayList<>();
        for (Skill skill : skills) {
            skillInfoList.add(new SkillInfo(skill.getId().toString(), user.getSkillLevel(skill), user.getSkillXp(skill)));
        }


        return UserSkillData.builder()
                .skills(skillInfoList)
                .mana(user.getMana())
                .uuid(uuid)
                .timestamp(System.currentTimeMillis())
                .server(ARC.serverName)
                .build();
    }

    public void applySkillData(UserSkillData data) {
        UUID uuid = data.uuid;
        SkillsUser user = AuraSkillsApi.get().getUser(uuid);
        if(user == null){
            ARC.warn("User with uuid {} not found while applying skill data", uuid);
            return;
        }
        for (SkillInfo skillInfo : data.skills) {
            NamespacedId id = NamespacedId.fromString(skillInfo.id);
            Skill skill = AuraSkillsApi.get().getGlobalRegistry().getSkill(id);
            if (skill == null) {
                ARC.warn("Skill with id {} not found", skillInfo.id);
                continue;
            }
            user.setSkillLevel(skill, skillInfo.level);
            user.setSkillXp(skill, skillInfo.xp);
        }
        user.setMana(data.mana);
    }

    @AllArgsConstructor
    @NoArgsConstructor
    @Data
    @Builder
    public static class UserSkillData implements SyncData {
        @SerializedName("u")
        UUID uuid;
        @SerializedName("sk")
        List<SkillInfo> skills;
        @SerializedName("m")
        double mana;
        @SerializedName("t")
        long timestamp;
        @SerializedName("s")
        String server;

        @Override
        public long timestamp() {return timestamp;}

        @Override
        public String server() {return server;}

        @Override
        public UUID uuid() {return uuid;}
    }


}

File: src/main/kotlin/ru/arc/sync/SkillsSync.kt
package ru.arc.sync

import com.google.gson.annotations.SerializedName
import dev.aurelium.auraskills.api.AuraSkillsApi
import dev.aurelium.auraskills.api.registry.NamespacedId
import org.bukkit.Bukkit
import ru.arc.ARC
import ru.arc.core.ScheduledTask
import ru.arc.core.repeating
import ru.arc.core.ticks
import ru.arc.sync.base.Context
import ru.arc.sync.base.Sync
import ru.arc.sync.base.SyncData
import ru.arc.sync.base.SyncRepo
import ru.arc.util.Logging
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicInteger

class SkillsSync : Sync {

    private val syncRepo =
        SyncRepo(
            clazz = UserSkillData::class.java,
            key = "arc.skills_data",
            redisManager = checkNotNull(ARC.redisManager) { "Redis manager is not initialized" },
            dataApplier = ::applySkillData,
            dataProducer = ::getSkillData,
        )

    private val loaded: MutableMap<UUID, Boolean> = ConcurrentHashMap()
    private val joinTasks = ConcurrentHashMap<UUID, ScheduledTask>()

    override fun playerJoin(uuid: UUID) {
        val counter = AtomicInteger(0)
        val task =
            repeating(5.ticks, delay = 5.ticks) {
                if (Bukkit.getPlayer(uuid) == null) {
                    cancelJoinTask(uuid)
                    return@repeating
                }
                val user = AuraSkillsApi.get().getUser(uuid)
                if (user == null) {
                    if (counter.incrementAndGet() > 20) {
                        Logging.warn("SkillsUser is null for {} for 20 cycles. Cancelling task.", uuid)
                        cancelJoinTask(uuid)
                    }
                    return@repeating
                }
                syncRepo
                    .loadAndApplyData(uuid)
                    .whenComplete { _, failure ->
                        if (failure == null && Bukkit.getPlayer(uuid) != null) {
                            loaded[uuid] = true
                        } else {
                            loaded.remove(uuid)
                        }
                    }
                cancelJoinTask(uuid)
            }
        joinTasks.put(uuid, task)?.cancel()
    }

    override fun forceSave(uuid: UUID) {
        if (loaded[uuid] != true) return
        val context = Context()
        context.put("uuid", uuid)
        syncRepo.saveAndPersistData(context)
    }

    override fun playerQuit(uuid: UUID) {
        forceSave(uuid)
        loaded.remove(uuid)
        cancelJoinTask(uuid)
    }

    override fun shutdown() {
        joinTasks.values.forEach(ScheduledTask::cancel)
        joinTasks.clear()
        loaded.clear()
    }

    private fun cancelJoinTask(uuid: UUID) {
        joinTasks.remove(uuid)?.cancel()
    }

    fun getSkillData(context: Context): UserSkillData? {
        val uuid: UUID = context.get("uuid")
        val user = AuraSkillsApi.get().getUser(uuid)
        if (user == null) {
            Logging.warn("User with uuid {} not found while getting skill data", uuid)
            return null
        }

        val skills = AuraSkillsApi.get().globalRegistry.skills
        val skillInfoList = skills.map { skill ->
            SkillInfo(id = skill.id.toString(), level = user.getSkillLevel(skill), xp = user.getSkillXp(skill))
        }

        return UserSkillData(
            id = uuid,
            skills = skillInfoList,
            mana = user.mana,
            ts = System.currentTimeMillis(),
            srv = ARC.serverName ?: "",
        )
    }

    fun applySkillData(data: UserSkillData) {
        val uuid = data.id ?: return
        val user = AuraSkillsApi.get().getUser(uuid)
        if (user == null) {
            Logging.warn("User with uuid {} not found while applying skill data", uuid)
            return
        }

        data.skills.forEach { skillInfo ->
            val id = NamespacedId.fromString(skillInfo.id)
            val skill = AuraSkillsApi.get().globalRegistry.getSkill(id)
            if (skill == null) {
                Logging.warn("Skill with id {} not found", skillInfo.id)
                return@forEach
            }
            if (user.getSkillLevel(skill) != skillInfo.level) user.setSkillLevel(skill, skillInfo.level)
            if (user.getSkillXp(skill) != skillInfo.xp) user.setSkillXp(skill, skillInfo.xp)
        }

        if (user.mana != data.mana) user.mana = data.mana
    }

    data class SkillInfo(
        @SerializedName("i") val id: String = "",
        @SerializedName("l") val level: Int = 0,
        @SerializedName("x") val xp: Double = 0.0,
    )

    data class UserSkillData(
        @SerializedName("u") val id: UUID? = null,
        @SerializedName("sk") val skills: List<SkillInfo> = emptyList(),
        @SerializedName("m") val mana: Double = 0.0,
        @SerializedName("t") val ts: Long = 0L,
        @SerializedName("s") val srv: String = "",
    ) : SyncData {
        override fun timestamp(): Long = ts
        override fun server(): String = srv
        override fun uuid(): UUID? = id
    }
}
